File: server/src/hephaestus/agent_bridge/jobstore.py
"""``py.jobstore_*`` durable store backing the thread-phase ``JobStore`` adapter.

The packaged sidecar has **no native SQLite addon** (repo_conventions Stage S
item 4); its application-owned async ``JobStore`` (``agent/src/workflows/
jobstore.ts``) persists every job/event/checkpoint by calling the five frozen
bridge methods — ``py.jobstore_{get,put,list,delete,checkpoint}`` — which this
module answers over dedicated ``tp_``-prefixed tables in the opstore
``state.db`` (DESIGN.md ``jobstore.py``; digest §5).

Design:

* A generic durable key/value namespace (``tp_jobstore``) carries the TS side's
  serialized job records and durable event log — ``put``/``get``/``list``/
  ``delete`` map straight onto it, upsert-idempotent and ordered by insertion
  ``seq`` so replay is deterministic.
* ``checkpoint`` writes a resumable phase checkpoint (``tp_jobstore_checkpoints``)
  carrying the **workflow version and input/output hashes** (digest §5, arch
  §4.5) so a resumed run continues only from a *verified* checkpoint. Checkpoints
  are upsert-idempotent on ``(job_id, checkpoint_key)`` and survive restart.

Every mutation runs inside opstore's ``BEGIN IMMEDIATE`` transaction, so
concurrent bridge writers serialize correctly across the single ``state.db``
connection. The ``tp_`` tables are created idempotently on the shared connection
(opstore's fixed schema owns no workflow tables and must not be edited).
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Final, cast

from opstore.db import Database
from opstore.errors import NotFoundError
from opstore.types import Clock, JSONValue, SystemClock

__all__ = [
    "CheckpointRecord",
    "JobStore",
    "KvRecord",
]

_KV_TABLE = "tp_jobstore"
_CHECKPOINT_TABLE = "tp_jobstore_checkpoints"

_CREATE_KV: Final[str] = f"""
CREATE TABLE IF NOT EXISTS {_KV_TABLE}(
  seq INTEGER PRIMARY KEY AUTOINCREMENT,
  namespace TEXT NOT NULL,
  key TEXT NOT NULL,
  value TEXT NOT NULL,
  updated_at REAL NOT NULL,
  UNIQUE(namespace, key))
"""

_CREATE_CHECKPOINTS: Final[str] = f"""
CREATE TABLE IF NOT EXISTS {_CHECKPOINT_TABLE}(
  job_id TEXT NOT NULL,
  checkpoint_key TEXT NOT NULL,
  workflow_version TEXT NOT NULL,
  input_hash TEXT NOT NULL,
  output_hash TEXT NOT NULL,
  value TEXT NOT NULL,
  updated_at REAL NOT NULL,
  PRIMARY KEY(job_id, checkpoint_key))
"""


@dataclass(frozen=True, slots=True)
class KvRecord:
    """One durable key/value row."""

    namespace: str
    key: str
    value: JSONValue
    updated_at: float


@dataclass(frozen=True, slots=True)
class CheckpointRecord:
    """One resumable phase checkpoint with its provenance hashes."""

    job_id: str
    checkpoint_key: str
    workflow_version: str
    input_hash: str
    output_hash: str
    value: JSONValue
    updated_at: float


def _loads(text: str) -> JSONValue:
    return cast(JSONValue, json.loads(text))


class JobStore:
    """Durable backing store for the sidecar's async thread-phase ``JobStore``.

    Construct once per opstore ``state.db``; the ``tp_`` tables are ensured on the
    shared connection at construction. All methods are synchronous over the
    single connection; the async ``dispatch`` wrapper adapts them to the five
    ``py.jobstore_*`` bridge requests.
    """

    def __init__(self, db: Database, *, clock: Clock | None = None) -> None:
        self._db = db
        self._clock = clock or SystemClock()
        db.conn.execute(_CREATE_KV)
        db.conn.execute(_CREATE_CHECKPOINTS)

    # -- generic key/value (jobs + durable event log) --------------------------

    def put(self, namespace: str, key: str, value: JSONValue) -> None:
        """Upsert a durable value; idempotent on ``(namespace, key)``."""
        text = json.dumps(value, sort_keys=True, ensure_ascii=False)
        now = self._clock.now()
        with self._db.transaction() as conn:
            conn.execute(
                f"INSERT INTO {_KV_TABLE}(namespace, key, value, updated_at) "
                "VALUES(?, ?, ?, ?) ON CONFLICT(namespace, key) "
                "DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at",
                (namespace, key, text, now),
            )

    def get(self, namespace: str, key: str) -> JSONValue | None:
        """The stored value, or ``None`` if absent."""
        row = self._db.conn.execute(
            f"SELECT value FROM {_KV_TABLE} WHERE namespace = ? AND key = ?",
            (namespace, key),
        ).fetchone()
        return None if row is None else _loads(str(row["value"]))

    def list(
        self, namespace: str, *, prefix: str | None = None, limit: int | None = None
    ) -> list[KvRecord]:
        """Records in ``namespace`` (optionally key-prefixed), insertion-ordered."""
        sql = f"SELECT namespace, key, value, updated_at FROM {_KV_TABLE} WHERE namespace = ?"
        params: list[object] = [namespace]
        if prefix is not None:
            sql += " AND key LIKE ? ESCAPE '\\'"
            params.append(_like_prefix(prefix))
        sql += " ORDER BY seq"
        if limit is not None:
            sql += " LIMIT ?"
            params.append(int(limit))
        rows = self._db.conn.execute(sql, tuple(params)).fetchall()
        return [
            KvRecord(
                namespace=str(r["namespace"]),
                key=str(r["key"]),
                value=_loads(str(r["value"])),
                updated_at=float(r["updated_at"]),
            )
            for r in rows
        ]

    def delete(self, namespace: str, key: str) -> bool:
        """Remove a value; returns ``True`` iff a row existed."""
        with self._db.transaction() as conn:
            cur = conn.execute(
                f"DELETE FROM {_KV_TABLE} WHERE namespace = ? AND key = ?",
                (namespace, key),
            )
        return cur.rowcount > 0

    # -- resumable checkpoints -------------------------------------------------

    def checkpoint(
        self,
        job_id: str,
        checkpoint_key: str,
        *,
        workflow_version: str,
        input_hash: str,
        output_hash: str,
        value: JSONValue,
    ) -> CheckpointRecord:
        """Persist a resumable phase checkpoint; idempotent on ``(job, key)``.

        The workflow version and input/output hashes are stored so resume can
        continue only from a *verified* checkpoint (arch §4.5, mission rule 6).
        """
        text = json.dumps(value, sort_keys=True, ensure_ascii=False)
        now = self._clock.now()
        with self._db.transaction() as conn:
            conn.execute(
                f"INSERT INTO {_CHECKPOINT_TABLE}(job_id, checkpoint_key, workflow_version, "
                "input_hash, output_hash, value, updated_at) VALUES(?, ?, ?, ?, ?, ?, ?) "
                "ON CONFLICT(job_id, checkpoint_key) DO UPDATE SET "
                "workflow_version = excluded.workflow_version, input_hash = excluded.input_hash, "
                "output_hash = excluded.output_hash, value = excluded.value, "
                "updated_at = excluded.updated_at",
                (job_id, checkpoint_key, workflow_version, input_hash, output_hash, text, now),
            )
        return CheckpointRecord(
            job_id=job_id,
            checkpoint_key=checkpoint_key,
            workflow_version=workflow_version,
            input_hash=input_hash,
            output_hash=output_hash,
            value=value,
            updated_at=now,
        )

    def get_checkpoint(self, job_id: str, checkpoint_key: str) -> CheckpointRecord | None:
        """The stored checkpoint for ``(job_id, checkpoint_key)``, if any."""
        row = self._db.conn.execute(
            f"SELECT * FROM {_CHECKPOINT_TABLE} WHERE job_id = ? AND checkpoint_key = ?",
            (job_id, checkpoint_key),
        ).fetchone()
        if row is None:
            return None
        return CheckpointRecord(
            job_id=str(row["job_id"]),
            checkpoint_key=str(row["checkpoint_key"]),
            workflow_version=str(row["workflow_version"]),
            input_hash=str(row["input_hash"]),
            output_hash=str(row["output_hash"]),
            value=_loads(str(row["value"])),
            updated_at=float(row["updated_at"]),
        )

    def list_checkpoints(self, job_id: str) -> list[CheckpointRecord]:
        """All checkpoints for a job, ordered by ``checkpoint_key``."""
        rows = self._db.conn.execute(
            f"SELECT * FROM {_CHECKPOINT_TABLE} WHERE job_id = ? ORDER BY checkpoint_key",
            (job_id,),
        ).fetchall()
        return [
            CheckpointRecord(
                job_id=str(r["job_id"]),
                checkpoint_key=str(r["checkpoint_key"]),
                workflow_version=str(r["workflow_version"]),
                input_hash=str(r["input_hash"]),
                output_hash=str(r["output_hash"]),
                value=_loads(str(r["value"])),
                updated_at=float(r["updated_at"]),
            )
            for r in rows
        ]

    # -- async bridge adapter --------------------------------------------------

    async def dispatch(self, method: str, params: dict[str, Any]) -> Any:
        """Answer one ``py.jobstore_*`` bridge request.

        Recognized methods: ``py.jobstore_get``/``_put``/``_list``/``_delete``/
        ``_checkpoint``. Unknown methods raise :class:`~opstore.errors.NotFoundError`.
        """
        if method == "py.jobstore_put":
            self.put(str(params["namespace"]), str(params["key"]), params.get("value"))
            return {"ok": True}
        if method == "py.jobstore_get":
            return {"value": self.get(str(params["namespace"]), str(params["key"]))}
        if method == "py.jobstore_list":
            prefix = params.get("prefix")
            limit = params.get("limit")
            records = self.list(
                str(params["namespace"]),
                prefix=None if prefix is None else str(prefix),
                limit=None if limit is None else int(cast(int, limit)),
            )
            return {"items": [{"key": r.key, "value": r.value} for r in records]}
        if method == "py.jobstore_delete":
            return {"deleted": self.delete(str(params["namespace"]), str(params["key"]))}
        if method == "py.jobstore_checkpoint":
            record = self.checkpoint(
                str(params["job_id"]),
                str(params["checkpoint_key"]),
                workflow_version=str(params["workflow_version"]),
                input_hash=str(params["input_hash"]),
                output_hash=str(params["output_hash"]),
                value=params.get("value"),
            )
            return {"ok": True, "updated_at": record.updated_at}
        raise NotFoundError(f"unknown jobstore method {method!r}")


def _like_prefix(prefix: str) -> str:
    escaped = prefix.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
    return escaped + "%"

File: agent/src/workflows/jobstore.ts
// Thread-phase `JobStore` implemented entirely over the private Python bridge.
//
// The packaged sidecar ships NO native SQLite addon and never opens a database
// (repo_conventions Stage S disposition 4; DESIGN.md `workflows/jobstore.ts`):
// every read and write here becomes one `py.jobstore_*` request answered by
// `hephaestus.agent_bridge.jobstore.JobStore` over the opstore `state.db`
// (`tp_`-prefixed tables). Thread-phase's `SqliteJobStore` is never constructed.
//
// Types come from the `@autonome-research/thread-phase/session` subpath only —
// the root barrel eagerly loads the transitive `openai` SDK (Stage S) and is
// eslint-forbidden inside agent/. Every import here is `import type`, so this
// module contributes zero thread-phase runtime code.
//
// ## Durable layout (CROSS-LANGUAGE CONTRACT with agent_bridge/workflows.py)
//
//   namespace `tp:jobs`    key `<jobId>`                     -> StoredJob (JSON)
//   namespace `tp:events`  key `<jobId>#<eventId, 12 digits>`-> StoredEvent (JSON)
//   namespace `tp:meta`    key `event_seq`                    -> last event id
//   namespace `tp:checkpoints` key `<jobId>#<checkpointKey>`  -> WorkflowCheckpoint
//
// The `tp:checkpoints` rows MIRROR the normative `tp_jobstore_checkpoints` table
// written by `py.jobstore_checkpoint`; the bridge has no checkpoint *read*
// method, and resume must verify a checkpoint's workflow version and input hash
// before skipping a phase.
//
// Field names in the stored records are the camelCase names of thread-phase's
// `JobRecord`/`EventRecord`, with `Date` fields as ISO-8601 strings. Python reads
// the same rows to replay durable events and to project orphaned `RUNNING` jobs
// as `interrupted`; changing a namespace, key format, or field name is a wire
// break on both sides.
//
// ## Atomicity
//
// The five bridge primitives are get/put/list/delete/checkpoint — there is no
// conditional write. First-writer-wins terminal transitions and
// `acquireExclusive` are therefore implemented as read-modify-write sequences
// serialized through an in-process mutex (`#serialize`). That is sufficient
// because exactly one workflow-runner process owns a job at a time: the Python
// `WorkflowService` supervises a single runner and owns cross-process
// exclusivity (leases + admission). Two concurrent runner processes writing the
// same job id would race — a documented limitation, not a supported mode.

import { randomUUID } from "node:crypto";
import type { JsonValue } from "../framing.js";
import type {
  EventRecord,
  GetJobOptions,
  JobFinalization,
  JobOwnership,
  JobRecord,
  JobStatus,
  JobStore,
  ListJobsOptions,
} from "@autonome-research/thread-phase/session";

/** One durable pipeline event (thread-phase's `PipelineEvent`, structurally). */
export type WorkflowEvent = Parameters<JobStore["appendEvent"]>[1];

/** Statuses that are actually persisted (`STALE` is computed at read time). */
export type PersistedJobStatus = Exclude<JobStatus, "STALE">;

/** Minimal bridge request surface (`RpcPeer.request`). */
export type BridgeCall = (
  method: string,
  params: { [k: string]: JsonValue },
) => Promise<JsonValue>;

/** Durable job row as stored by the bridge (ISO dates, JSON-safe values). */
export interface StoredJob {
  id: string;
  name: string;
  input: JsonValue;
  status: PersistedJobStatus;
  result: JsonValue;
  error: string | null;
  eventCount: number;
  createdAt: string;
  startedAt: string | null;
  completedAt: string | null;
  sessionId?: string;
  pid?: number;
  ppid?: number;
  cwd?: string;
  hostname?: string;
  ownerId?: string;
  launchSource?: string;
  heartbeatEnabled?: boolean;
  heartbeatAt?: string;
  /** Set by Python when a terminal is synthesized (e.g. `"interrupted"`). */
  failureClass?: string;
}

/** Durable event row as stored by the bridge. */
export interface StoredEvent {
  id: number;
  jobId: string;
  eventType: string;
  data: JsonValue;
  createdAt: string;
}

/** Resumable phase checkpoint (workflow version + input/output hashes). */
export interface WorkflowCheckpoint {
  readonly jobId: string;
  readonly checkpointKey: string;
  readonly workflowVersion: string;
  readonly inputHash: string;
  readonly outputHash: string;
  readonly value: JsonValue;
}

export interface BridgeJobStoreOptions {
  /** Namespace prefix; defaults to `tp` (the cross-language contract). */
  readonly namespacePrefix?: string;
  /** Job-id factory (injected in tests for deterministic ids). */
  readonly newJobId?: () => string;
  /** Clock (injected in tests). */
  readonly now?: () => Date;
}

const DEFAULT_LIST_LIMIT = 50;
const EVENT_KEY_DIGITS = 12;
const TERMINAL_STATUSES: ReadonlySet<PersistedJobStatus> = new Set<PersistedJobStatus>([
  "COMPLETED",
  "FAILED",
  "CANCELLED",
  "ABANDONED",
]);

/** `tp:events` key for one event id — zero-padded so prefix listing sorts. */
export function eventKey(jobId: string, eventId: number): string {
  return `${jobId}#${String(eventId).padStart(EVENT_KEY_DIGITS, "0")}`;
}

/** `tp:checkpoints` key for one phase checkpoint. */
export function checkpointKey(jobId: string, key: string): string {
  return `${jobId}#${key}`;
}

function toJson(value: unknown): JsonValue {
  if (value === undefined) return null;
  return JSON.parse(JSON.stringify(value)) as JsonValue;
}

function isRecord(value: JsonValue | null): value is { [k: string]: JsonValue } {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function isLive(status: PersistedJobStatus): boolean {
  return !TERMINAL_STATUSES.has(status);
}

/** Rehydrate the public `JobRecord` (Date fields) from its stored form. */
export function toJobRecord(stored: StoredJob, status: JobStatus = stored.status): JobRecord {
  const record: JobRecord = {
    id: stored.id,
    name: stored.name,
    input: stored.input,
    status,
    result: stored.result,
    error: stored.error,
    eventCount: stored.eventCount,
    createdAt: new Date(stored.createdAt),
    startedAt: stored.startedAt === null ? null : new Date(stored.startedAt),
    completedAt: stored.completedAt === null ? null : new Date(stored.completedAt),
    ...(stored.sessionId !== undefined ? { sessionId: stored.sessionId } : {}),
    ...(stored.pid !== undefined ? { pid: stored.pid } : {}),
    ...(stored.ppid !== undefined ? { ppid: stored.ppid } : {}),
    ...(stored.cwd !== undefined ? { cwd: stored.cwd } : {}),
    ...(stored.hostname !== undefined ? { hostname: stored.hostname } : {}),
    ...(stored.ownerId !== undefined ? { ownerId: stored.ownerId } : {}),
    ...(stored.launchSource !== undefined ? { launchSource: stored.launchSource } : {}),
    ...(stored.heartbeatEnabled !== undefined
      ? { heartbeatEnabled: stored.heartbeatEnabled }
      : {}),
    ...(stored.heartbeatAt !== undefined ? { heartbeatAt: new Date(stored.heartbeatAt) } : {}),
  };
  return record;
}

function toEventRecord(stored: StoredEvent): EventRecord {
  return {
    id: stored.id,
    jobId: stored.jobId,
    eventType: stored.eventType,
    data: stored.data as unknown as WorkflowEvent,
    createdAt: new Date(stored.createdAt),
  };
}

/**
 * Thread-phase `JobStore` whose every operation is a `py.jobstore_*` bridge
 * call. Construct with the sidecar's `RpcPeer.request`.
 */
export class BridgeJobStore implements JobStore {
  readonly #call: BridgeCall;
  readonly #jobsNs: string;
  readonly #eventsNs: string;
  readonly #metaNs: string;
  readonly #checkpointsNs: string;
  readonly #newJobId: () => string;
  readonly #now: () => Date;
  #tail: Promise<unknown> = Promise.resolve();
  #nextEventId: number | undefined;

  constructor(call: BridgeCall, options: BridgeJobStoreOptions = {}) {
    const prefix = options.namespacePrefix ?? "tp";
    this.#call = call;
    this.#jobsNs = `${prefix}:jobs`;
    this.#eventsNs = `${prefix}:events`;
    this.#metaNs = `${prefix}:meta`;
    this.#checkpointsNs = `${prefix}:checkpoints`;
    this.#newJobId = options.newJobId ?? (() => `job-${randomUUID()}`);
    this.#now = options.now ?? (() => new Date());
  }

  // -- bridge primitives ----------------------------------------------------

  async #get(namespace: string, key: string): Promise<JsonValue | null> {
    const raw = await this.#call("py.jobstore_get", { namespace, key });
    if (!isRecord(raw)) return null;
    const value = raw.value;
    return value === undefined ? null : value;
  }

  async #put(namespace: string, key: string, value: JsonValue): Promise<void> {
    await this.#call("py.jobstore_put", { namespace, key, value });
  }

  async #delete(namespace: string, key: string): Promise<boolean> {
    const raw = await this.#call("py.jobstore_delete", { namespace, key });
    return isRecord(raw) && raw.deleted === true;
  }

  async #list(
    namespace: string,
    prefix?: string,
    limit?: number,
  ): Promise<Array<{ key: string; value: JsonValue }>> {
    const params: { [k: string]: JsonValue } = { namespace };
    if (prefix !== undefined) params.prefix = prefix;
    if (limit !== undefined) params.limit = limit;
    const raw = await this.#call("py.jobstore_list", params);
    if (!isRecord(raw) || !Array.isArray(raw.items)) return [];
    const items: Array<{ key: string; value: JsonValue }> = [];
    for (const entry of raw.items) {
      if (!isRecord(entry) || typeof entry.key !== "string") continue;
      items.push({ key: entry.key, value: entry.value ?? null });
    }
    return items;
  }

  /** Serialize read-modify-write sequences so they are atomic in this process. */
  #serialize<T>(body: () => Promise<T>): Promise<T> {
    const run = this.#tail.then(body, body);
    this.#tail = run.then(
      () => undefined,
      () => undefined,
    );
    return run;
  }

  // -- job records ----------------------------------------------------------

  async #readJob(jobId: string): Promise<StoredJob | null> {
    const raw = await this.#get(this.#jobsNs, jobId);
    return isRecord(raw) ? (raw as unknown as StoredJob) : null;
  }

  async #writeJob(job: StoredJob): Promise<void> {
    await this.#put(this.#jobsNs, job.id, job as unknown as JsonValue);
  }

  /** The stored (pre-rehydration) row — used by workflow resume and tests. */
  storedJob(jobId: string): Promise<StoredJob | null> {
    return this.#readJob(jobId);
  }

  async createJob(name: string, input: unknown): Promise<string> {
    return this.createJobWithId(this.#newJobId(), name, input);
  }

  /**
   * `createJob` with a caller-supplied id.
   *
   * The Python `WorkflowService` mints the id before launching so it can cancel
   * or inspect the run while `workflow.run` is still in flight. Re-creating an
   * existing id is refused — a job row is never silently reset.
   */
  async createJobWithId(id: string, name: string, input: unknown): Promise<string> {
    return this.#serialize(async () => {
      if ((await this.#readJob(id)) !== null) throw new Error(`job ${id} already exists`);
      const job: StoredJob = {
        id,
        name,
        input: toJson(input),
        status: "PENDING",
        result: null,
        error: null,
        eventCount: 0,
        createdAt: this.#now().toISOString(),
        startedAt: null,
        completedAt: null,
      };
      await this.#writeJob(job);
      return id;
    });
  }

  async acquireExclusive(name: string, input: unknown): Promise<string | null> {
    return this.#serialize(async () => {
      for (const entry of await this.#list(this.#jobsNs)) {
        if (!isRecord(entry.value)) continue;
        const job = entry.value as unknown as StoredJob;
        if (job.name === name && job.status === "RUNNING") return null;
      }
      const id = this.#newJobId();
      const startedAt = this.#now().toISOString();
      const job: StoredJob = {
        id,
        name,
        input: toJson(input),
        status: "RUNNING",
        result: null,
        error: null,
        eventCount: 0,
        createdAt: startedAt,
        startedAt,
        completedAt: null,
      };
      await this.#writeJob(job);
      return id;
    });
  }

  async setRunning(jobId: string, ownership?: JobOwnership): Promise<boolean> {
    return this.#serialize(async () => {
      const job = await this.#readJob(jobId);
      if (job === null) return false;
      if (job.status !== "PENDING" && job.status !== "RUNNING") return false;
      const claimedOwner = ownership?.ownerId;
      if (job.ownerId !== undefined && claimedOwner !== undefined && job.ownerId !== claimedOwner) {
        return false;
      }
      const next: StoredJob = { ...job, status: "RUNNING" };
      if (ownership !== undefined) applyOwnership(next, ownership);
      next.startedAt = job.startedAt ?? this.#now().toISOString();
      await this.#writeJob(next);
      return true;
    });
  }

  /** Shared terminal CAS: only a live row owned by `ownerId` may transition. */
  async #finalize(
    jobId: string,
    status: PersistedJobStatus,
    patch: { result?: JsonValue; error?: string | null },
    ownerId?: string,
  ): Promise<StoredJob | null> {
    const job = await this.#readJob(jobId);
    if (job === null || !isLive(job.status)) return null;
    if (ownerId !== undefined && job.ownerId !== undefined && job.ownerId !== ownerId) return null;
    const next: StoredJob = {
      ...job,
      status,
      completedAt: this.#now().toISOString(),
      ...(patch.result !== undefined ? { result: patch.result } : {}),
      ...(patch.error !== undefined ? { error: patch.error } : {}),
    };
    await this.#writeJob(next);
    return next;
  }

  setCompleted(jobId: string, result: unknown, ownerId?: string): Promise<boolean> {
    return this.#serialize(async () =>
      (await this.#finalize(jobId, "COMPLETED", { result: toJson(result) }, ownerId)) !== null,
    );
  }

  setFailed(jobId: string, error: string, ownerId?: string): Promise<boolean> {
    return this.#serialize(async () =>
      (await this.#finalize(jobId, "FAILED", { error }, ownerId)) !== null,
    );
  }

  setCancelled(jobId: string, reason: string, ownerId?: string): Promise<boolean> {
    return this.#serialize(async () =>
      (await this.#finalize(jobId, "CANCELLED", { error: reason }, ownerId)) !== null,
    );
  }

  setAbandoned(jobId: string, reason: string): Promise<boolean> {
    return this.#serialize(async () =>
      (await this.#finalize(jobId, "ABANDONED", { error: reason })) !== null,
    );
  }

  setAbandonedIfStale(
    jobId: string,
    staleBefore: Date,
    reason: string,
    expectedOwnerId?: string,
  ): Promise<boolean> {
    return this.#serialize(
      async () => (await this.#abandonIfStale(jobId, staleBefore, reason, expectedOwnerId)) !== null,
    );
  }

  async #abandonIfStale(
    jobId: string,
    staleBefore: Date,
    reason: string,
    expectedOwnerId?: string,
  ): Promise<StoredJob | null> {
    const job = await this.#readJob(jobId);
    if (job === null || job.status !== "RUNNING") return null;
    if (job.heartbeatAt !== undefined && new Date(job.heartbeatAt) >= staleBefore) return null;
    return this.#finalize(jobId, "ABANDONED", { error: reason }, expectedOwnerId);
  }

  finalizeJob(jobId: string, finalization: JobFinalization): Promise<EventRecord | null> {
    return this.#serialize(async () => {
      const patch: { result?: JsonValue; error?: string | null } =
        finalization.status === "COMPLETED"
          ? { result: toJson(finalization.result) }
          : { error: finalization.error ?? "" };
      const job = await this.#finalize(jobId, finalization.status, patch, finalization.ownerId);
      if (job === null) return null;
      return this.#appendEvent(job, finalization.event);
    });
  }

  finalizeAbandonedIfStale(
    jobId: string,
    staleBefore: Date,
    reason: string,
    expectedOwnerId?: string,
  ): Promise<EventRecord | null> {
    return this.#serialize(async () => {
      const job = await this.#abandonIfStale(jobId, staleBefore, reason, expectedOwnerId);
      if (job === null) return null;
      return this.#appendEvent(job, { type: "error", message: reason } as WorkflowEvent);
    });
  }

  async heartbeat(jobId: string, ownerId?: string): Promise<void> {
    await this.#serialize(async () => {
      const job = await this.#readJob(jobId);
      if (job === null || job.status !== "RUNNING") return;
      if (ownerId !== undefined && job.ownerId !== undefined && job.ownerId !== ownerId) return;
      await this.#writeJob({ ...job, heartbeatAt: this.#now().toISOString() });
    });
  }

  enableHeartbeat(jobId: string, ownerId: string): Promise<boolean> {
    return this.#serialize(async () => {
      const job = await this.#readJob(jobId);
      if (job === null || job.status !== "RUNNING") return false;
      if (job.ownerId !== undefined && job.ownerId !== ownerId) return false;
      await this.#writeJob({
        ...job,
        ownerId,
        heartbeatEnabled: true,
        heartbeatAt: this.#now().toISOString(),
      });
      return true;
    });
  }

  async getJob(jobId: string, options?: GetJobOptions): Promise<JobRecord | null> {
    const job = await this.#readJob(jobId);
    if (job === null) return null;
    return toJobRecord(job, this.#readStatus(job, options?.staleAfterMs));
  }

  /** `STALE` is never persisted: it is computed from `heartbeatAt` on read. */
  #readStatus(job: StoredJob, staleAfterMs?: number): JobStatus {
    if (staleAfterMs === undefined || job.status !== "RUNNING") return job.status;
    if (job.heartbeatAt === undefined) return job.status;
    const cutoff = this.#now().getTime() - staleAfterMs;
    return new Date(job.heartbeatAt).getTime() < cutoff ? "STALE" : job.status;
  }

  async listJobs(options: ListJobsOptions = {}): Promise<JobRecord[]> {
    const entries = await this.#list(this.#jobsNs);
    const records: JobRecord[] = [];
    for (const entry of entries) {
      if (!isRecord(entry.value)) continue;
      const job = entry.value as unknown as StoredJob;
      if (options.name !== undefined && job.name !== options.name) continue;
      const status = this.#readStatus(job, options.staleAfterMs);
      if (options.status !== undefined && status !== options.status) continue;
      records.push(toJobRecord(job, status));
    }
    // Newest first, matching the bundled SqliteJobStore's `created_at DESC`.
    records.sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
    return records.slice(0, options.limit ?? DEFAULT_LIST_LIMIT);
  }

  // -- durable event log ----------------------------------------------------

  async #reserveEventId(): Promise<number> {
    if (this.#nextEventId === undefined) {
      const raw = await this.#get(this.#metaNs, "event_seq");
      this.#nextEventId = typeof raw === "number" ? raw + 1 : 1;
    }
    const id = this.#nextEventId;
    this.#nextEventId = id + 1;
    await this.#put(this.#metaNs, "event_seq", id);
    return id;
  }

  /** Append + bump the job's `eventCount` (caller already holds the mutex). */
  async #appendEvent(job: StoredJob, event: WorkflowEvent): Promise<EventRecord> {
    const id = await this.#reserveEventId();
    const stored: StoredEvent = {
      id,
      jobId: job.id,
      eventType: String((event as { type?: unknown }).type ?? "unknown"),
      data: toJson(event),
      createdAt: this.#now().toISOString(),
    };
    await this.#put(this.#eventsNs, eventKey(job.id, id), stored as unknown as JsonValue);
    await this.#writeJob({ ...job, eventCount: job.eventCount + 1 });
    return toEventRecord(stored);
  }

  appendEvent(jobId: string, event: WorkflowEvent): Promise<number> {
    return this.#serialize(async () => {
      const job = await this.#readJob(jobId);
      if (job === null) throw new Error(`unknown job ${jobId}`);
      const record = await this.#appendEvent(job, event);
      return record.id;
    });
  }

  async getEvents(jobId: string, afterId = 0): Promise<EventRecord[]> {
    const entries = await this.#list(this.#eventsNs, `${jobId}#`);
    const records: EventRecord[] = [];
    for (const entry of entries) {
      if (!isRecord(entry.value)) continue;
      const stored = entry.value as unknown as StoredEvent;
      if (stored.id <= afterId) continue;
      records.push(toEventRecord(stored));
    }
    records.sort((a, b) => a.id - b.id);
    return records;
  }

  // -- resumable checkpoints (beyond the JobStore interface) ---------------

  /**
   * Persist one resumable phase checkpoint with its workflow version and
   * input/output hashes (`py.jobstore_checkpoint`; digest §5).
   *
   * Two writes, deliberately: `py.jobstore_checkpoint` fills the normative
   * `tp_jobstore_checkpoints` table (what Python verifies and what survives as
   * the durable record), and a `tp:checkpoints` key/value mirror gives the
   * sidecar a read path — the bridge exposes no checkpoint *read* method, and
   * resume must verify version/input hashes before skipping a phase.
   */
  async checkpoint(record: WorkflowCheckpoint): Promise<void> {
    await this.#call("py.jobstore_checkpoint", {
      job_id: record.jobId,
      checkpoint_key: record.checkpointKey,
      workflow_version: record.workflowVersion,
      input_hash: record.inputHash,
      output_hash: record.outputHash,
      value: record.value,
    });
    await this.#put(
      this.#checkpointsNs,
      checkpointKey(record.jobId, record.checkpointKey),
      record as unknown as JsonValue,
    );
  }

  /** The mirrored checkpoint for `(jobId, checkpointKey)`, if one was written. */
  async readCheckpoint(jobId: string, key: string): Promise<WorkflowCheckpoint | null> {
    const raw = await this.#get(this.#checkpointsNs, checkpointKey(jobId, key));
    return isRecord(raw) ? (raw as unknown as WorkflowCheckpoint) : null;
  }

  /** Every mirrored checkpoint for a job, keyed by `checkpointKey`. */
  async listCheckpoints(jobId: string): Promise<Map<string, WorkflowCheckpoint>> {
    const entries = await this.#list(this.#checkpointsNs, `${jobId}#`);
    const found = new Map<string, WorkflowCheckpoint>();
    for (const entry of entries) {
      if (!isRecord(entry.value)) continue;
      const record = entry.value as unknown as WorkflowCheckpoint;
      found.set(record.checkpointKey, record);
    }
    return found;
  }

  /** Drop a job row (its events stay unless removed explicitly). */
  deleteJob(jobId: string): Promise<boolean> {
    return this.#serialize(() => this.#delete(this.#jobsNs, jobId));
  }

  /** No-op: the durable connection lives in Python, not in the sidecar. */
  close(): void {}
}

function applyOwnership(job: StoredJob, ownership: JobOwnership): void {
  if (ownership.sessionId !== undefined) job.sessionId = ownership.sessionId;
  if (ownership.pid !== undefined) job.pid = ownership.pid;
  if (ownership.ppid !== undefined) job.ppid = ownership.ppid;
  if (ownership.cwd !== undefined) job.cwd = ownership.cwd;
  if (ownership.hostname !== undefined) job.hostname = ownership.hostname;
  if (ownership.ownerId !== undefined) job.ownerId = ownership.ownerId;
  if (ownership.launchSource !== undefined) job.launchSource = ownership.launchSource;
  if (ownership.heartbeatEnabled !== undefined) job.heartbeatEnabled = ownership.heartbeatEnabled;
}
